src/index.ml

```diff
@@ -237,17 +237,17 @@ struct
         else if offset < h.offset then sync_log_entries ~min:offset log
         else if offset > h.offset then assert false
 
-  let sync_index ~generation t =
-    may (fun (i : index) -> IO.close i.io) t.index;
+  let sync_index t =
+    Option.iter (fun (i : index) -> IO.close i.io) t.index;
     let index_path = Layout.data ~root:t.root in
-    if not (IO.exists index_path) then t.index <- None
-    else
+    if IO.exists index_path then
       let io =
-        IO.v ~fresh:false ~readonly:true ~generation ~fan_size:0L index_path
+        IO.v ~fresh:false ~readonly:true ~generation:0L ~fan_size:0L index_path
```

Comment:
Might warrant a comment saying that this argument isn't used in this combination.